NG-84 - nested and array fields (#5)
nerdgeschoss/react-use-form-library: Readme.md

---

## Nested Objects

> Each FormField object is also able to contain fields. It behaves similar as with the Form object but with slight differences on it's methods (See [API](#nested-fields)).

Reviewer: ```suggestion
Each FormField object is also able to contain fields. It behaves similarly to the Form object but with slight differences in its methods (See [API](#nested-fields)).
```

```ts
import { useForm } from 'react-use-form-library';

function App(): JSX.Element {
  const { model, fields, onSubmit } = useForm({
    model: {
      address: {
        streetName: '',
      },
    },
    handleSubmit: async () => {
      console.log(model);
    },
  });

  return (
    <form onSubmit={onSubmit}>
      <input

>         /* Since the FormField could be either a value or an object, the fields property is optional 

Reviewer: I don't really get this restriction. From the type system it should be clear at this point that this field is a required object.

Author: You're right. This will indeed give you a typescript error based on your model type.

        and you will need to check for existance */
        value={fields.address.fields?.streetName.value}
        onChange={(v) =>
          fields.address.fields?.streetName.onChange(v.target.value)
        }
      />
      <button>Submit</button>
    </form>
  );
}
```

Validations also work with nested objects

```ts
import { useForm } from 'react-use-form-library';

function App(): JSX.Element {
  const { model, fields, onSubmit } = useForm({
    model: {
      address: {
        streetName: '',
      },
    },
    handleSubmit: async () => {
      console.log(model);
    },
    validations: {
      address: {
        streetName: 'required',
      },
    }
  });

  // ...
```

Since the fields property within FormField is also based on a proxy object, you can access nested properties without
explicitely declaring them

```ts
function App(): JSX.Element {
  const { model, fields, onSubmit } = useForm({
    model: {},
    handleSubmit: async () => {
      console.log(model);
    },
  });

  return (
    <form onSubmit={onSubmit}>
      <input

>         value={fields.bankDetails.fields?.address.fields?.streetName.value}

Reviewer: this shouldn't be valid typescript here - the model type is not declared anywhere, so this would not prevent you from typos anymore when dealing with forms.

Author: Yes this will result in a typescript error if you don't supply the interface. I'll update the example.

        onChange={(v) =>
          fields.bankDetails.fields?.address.fields?.streetName.onChange(
            v.target.value
          )
        }
      />
      <button>Submit</button>
    </form>
  );
}
```

---

## FieldSet

> A FieldSet is basically an array of FormFields. Contrary to nested objects, a FieldSet will only be instantiated if it is **explicitely** declared in the model.

Reviewer: ```suggestion
A FieldSet is basically an array of FormFields. Contrary to nested objects, a FieldSet will only be instantiated if it is **explicitly** declared in the model.
```

> This is because at the moment of field instantiation, the addField method within a Form (or a FormField if it is a nested object) will as if the model value is actually an Array. It is not possible of implicitely determine if the field is an array based only on the type.

Reviewer: ```suggestion
This is because at the moment of field instantiation, the `addField` method within a `Form` (or a `FormField` if it is a nested object) will as if the model value is actually an Array. It is not possible to implicitly determine if the field is array-based only on the type.
```

The FieldSet class extends Array, so you can access all methods an array has, plus custom methods that work similarly as those in FormField (more in the [API](#fieldset-1))

```ts
function App(): JSX.Element {
  const { model, fields, onSubmit } = useForm({
    model: {
      emails: ['example@email.com'],
    },
    handleSubmit: async () => {
      console.log(model);
    },
  });

  return (
    <form onSubmit={onSubmit}>
      // a FieldSet is iterable
      {fields.emails.map((field, index) => {
        // Every FieldSet item is a FormField
        return (
          <div className="input">
            <input
              value={field.value}
              onChange={(v) => field.onChange(v.target.value)}
            />
            // FieldSet has a helpful removeField method

>             <button onClick={() => fields.emails.remove(index)}>remove</button>

Reviewer: why do we even need the index here? the `field` object should know its index, so `field.remove()` is a lot easier to call.

Author: Is there a more obvious way of achieven this? My first guess was to provide a function on field creation like this:

```
items.forEach((item) => {
      this.fields.push(
        new FormField({
          value: item,
          onUpdate: this.onUpdate,
          validation: this.validation,
          remove: () => {
            this.fields.splice(Math.max(this.fields.length - 1, 0), 1);
          },
        })
      );
    });
```

Author: added remove method

          </div>
        );
      })}
      // FieldSet also has an addFields method to create new items
      <button onClick={() => fields.emails.addFields('')}>Add Field</button>
      <button>Submit</button>
    </form>
  );
}
```

### Adding Items

> The `addFields` methods takes an array of items that can be either a simple value or an object with value and validation. Being an array extension, it is also possible to add items to FieldSet using array methods like `push`;

Reviewer: I'm not sure if you're doing yourself a favor by extending `Array`. While in theory it makes sense, I haven't yet witnessed a single case where it turned out to be a good idea in the end.

Reviewer: If instead you return a `fields`-Array from the property, it would also be closer to the API of nested objects:
```ts
fields.emails.fields.each
fields.emails.add('newmail')
```

```ts
// simple value
<button onClick={() => fields.emails.addFields('example@email.com')}>Add Field</button>
// value with validation

> <button onClick={() => fields.emails.addFields({

Reviewer: why should different objects in the same array have different validations?

Author: Removed this posibility, now the validation applies to all items

  value: '',
  validation: 'email'
})}>Add Field</button>
// Multiple Fields
<button onClick={() => fields.emails.addFields('example@email.com', '', 'another-example@emal.com')}>Add Field</button>
```

**Important**

> `addFields` takes a comma separated array of parameters, if you would like to pass an array you will need to destructure it

Reviewer: this is something you can mitigate in the implementation and just accept both if it is important

```ts
const newValues = ['example@email.com', '', 'another-example@emal.com'];
<button onClick={() => fields.emails.addFields(...newValues)}
```

### Removing Items

There is a `removeField` helper function that takes in an index number as parameter. Under the hood it is a simple call to `splice`. You can also use array methods like `pop` or `shift` to remove items.

```ts
<button onClick={() => fields.emails.remove(index)}>remove</button>
```

> ---

Reviewer: What I'm missing here is the main use case of arrays: arrays of objects.
Assuming I have a model like 
```ts
interface Model {
  images: Array<{
    id: string;
    tags: string[];
    file?: Blob;
  }>
}
```

I want to access it like
```ts
form.fields.images.fields[0].tags.valid;
form.fields.images.fields[0].file.required;
form.fields.images.add({id: "1", tags: []});
```
